### ISSUE TITLE
training with reset-position-ids and reset-attention-mask

### ISSUE THREAD

**Issue by toothacher17** [HazyResearch/flash-attention#161]
Hi, I am trying to use flash-attention in megatron and I am wondering if I am pretraining with reset-position-ids and reset-attention-mask, how should I pass the customized block-wise diagonal attention-masks to use flash-attention?

For example, without reset attention mask, the attention mask matrix will be:
```
0. -inf -inf -inf -inf
0  0     -inf -inf -inf
0  0     0.    -inf -inf
0  0     0.    0     -inf
0  0     0.    0     0
```

as we are resetting attention mask (say one sentence ends at position 2), the attention matrix will become block-diagonal:
```
0    -inf   -inf   -inf  -inf
0     0     -inf   -inf  -inf
-inf -inf   0      -inf  -inf
-inf -inf   0      0     -inf
-inf -inf   0      0      0
```

For a typical input with batch size B, we will have B different block-diagonal attention masks (as each sequence has different resetting positions), what is the right API that I should use? From megatron-LM, the implementation is: 
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L393

```
        output = flash_attn_unpadded_func(
            q, k, v, cu_seqlens_q, cu_seqlens_k, seqlen_q, seqlen_k,
            self.dropout_p,
            softmax_scale=self.softmax_scale, causal=is_causal
        )
```

I believe this is not really correct, as it is using a lower triangular mask, but not really the case of reset-attention-mask that I was describing. 

Could you help me by pointing to the right API to use in megatron-LM or is this functionality implemented at all. I tried to use xformers but it will throw weird NANs when using cutlass based ops:
https://github.com/facebookresearch/xformers/issues/722

Thanks in advance! I personally think this should be related with the new features of support attention bias as described in your roadmap, e.g. block-diagonal is a pretty common customized attention bias. To implement attention bias like ALiBi, I believe your API will allow pass in a customized attention bias mask by any means?

**Comment by toothacher17** [comment]
Hi, this ticket can be closed. I implemented with the flash_attn_unpadded_func. As long as we calculated the cu_seqlens_q correctly, we can get the correct block diagonal matrix.

**Comment by whalefa1I** [comment]
> As long as we calculated the cu_seqlens_q correctly, we can get the correct block diagonal matrix.

I am using alibi, megatron-lm, and flash-attn. When I enable both reset-position-ids and reset-attention-mask at the same time, I can't get the correct loss. However, when I enable reset-attention-mask, the loss becomes normal. Could I know some details about your implementation process?

**Comment by toothacher17** [comment]
> However, when I enable reset-attention-mask, the loss becomes normal.

If you are using megatron-lm post-processing logic and flash attention at the same time, and your loss is not working as expected, you are probably getting something wrong. 

The post-processing logic with reset-attention-mask probably does not help with flash attention, as it does not use the attention mask at all. It used the cu-seqlens?

**Comment by toothacher17** [comment]
Are you using the customized attention mask to implement with your alibi? I heard some people are using the triton version of flash-attn v2 and it is working fine
